mlflow/telemetry/client.py

```diff
@@ -1,33 +1,251 @@
+            self._worker_threadpool = ThreadPoolExecutor(
+                max_workers=self._max_workers,
+                thread_name_prefix="MlflowTelemetryWorker",
+            )
+            self._consumer_thread = threading.Thread(
+                target=self._consumer_loop,
+                name="MLflowTelemetryConsumer",
+                daemon=True,
+            )
+            self._consumer_thread.start()
+
+    def _at_exit_callback(self) -> None:
+        """Callback function executed when the program is exiting."""
+        _logger.debug(
+            "Flushing the async telemetry queue before program exit. This may take a while..."
+        )
+        self.flush(terminate=True)
+
+    def flush(self, terminate=False) -> None:
+        """
+        Flush the async telemetry queue.
+
+        Args:
+            terminate: If True, shut down the telemetry threads after flushing.
+        """
+        if not self.is_active():
+            return
+
+        try:
+            # Send any pending records before flushing
+            with self._batch_lock:
+                if self._pending_records:
+                    self._send_batch()
+
+            self._stop_event.set()
+
+            try:
+                self._consumer_thread.join(timeout=30)
+            except Exception as e:
+                _logger.debug(f"Error waiting for consumer thread: {e}")
+
+            try:
+                self._queue.join()
+            except Exception as e:
+                _logger.debug(f"Error waiting for queue to drain: {e}")
+
+            try:
+                self._worker_threadpool.shutdown(wait=True, timeout=30)
+            except Exception as e:
+                _logger.debug(f"Error shutting down worker thread pool: {e}")
+
+            self._is_active = False
+
+            # Restart threads to listen to incoming requests after flushing, if not terminating
+            if not terminate:
+                self._stop_event.clear()
+                self.activate()
+
+        except Exception as e:
+            _logger.debug(f"Error during telemetry flush: {e}", exc_info=True)
+            # Ensure we mark as inactive even if there was an error
+            self._is_active = False
 
     def _update_backend_store(self):
         """
         Backend store might be changed after mlflow is imported, we should use this
         method to update the backend store info at sending telemetry step.
         """
+        # import here to avoid circular import
+        from mlflow.tracking._tracking_service.utils import _get_store
+
         tracking_store = _get_store()
         self.info.backend_store = tracking_store.__class__.__name__
 
+    def _generate_telemetry_record(self, record: APIRecord) -> TelemetryRecord:
+        self._update_backend_store()
+        telemetry_info = asdict(self.info)
+        # TODO: update partition key
+        return TelemetryRecord(
+            data=json.dumps(telemetry_info | asdict(record)), partition_key="test"
```

Comment:
why do we need `json.dumps` here?

Reply:
> why do we need json.dumps here?

Because data should a string.

> Any chance telemetry_info and record has intersects?

No, record is APIRecord format, telemetry_info is TelemetryInfo format.

Reply:
> Because data should a string.

What happens if we send `{"records": [{"data": {"foo": "bar"}}]}`?

Reply:
I think when it reaches the telemetry url it will fail with some validation error, we will set up that later. Kinesis PutRecords operation only accepts string data input, and this data field will be parsed as that field

Reply:
Got it, thanks for the explanation!